Relax `DGMulti` `<:Polynomial` type restriction

jlchan opened the pull request:
Currently, `DGMulti` specializes `rhs!` on the `approximation_type`. The default is `Polynomial`; however, the implementation of `rhs!` for `Polynomial` types should run for any `DGMulti` discretization built using StartUpDG.jl, though it may be slow (this is because StartUpDG.jl just represents an approximation using multi-dimensional matrix operators). 

This PR removes the `<:Polynomial` type restriction, and instead makes the `Polynomial` implementation of `rhs!` the default `DGMulti` implementation. This way, if we introduce a new approximation type in StartUpDG.jl, Trixi.jl should automatically be able to accommodate it.

jlchan commented:
@andrewwinters5000 I requested your review because you mentioned you were interested in understanding how the `DGMulti` solvers work. Feel free to disregard if this doesn't seem relevant to you.

ranocha commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -393,11 +393,12 @@ end
+# General fallback for DGMulti solvers:
 # Polynomial-based solvers use hybridized SBP operators, which have blocks scaled by outward
 # normal components. This implies that operators for different coordinate directions have
 # different sparsity patterns. We default to using sum factorization (which is faster when
 # operators are sparse) for all `<:Polynomial` approximation types.
```
```suggestion
# different sparsity patterns. We default to using sum factorization (which is faster when
# operators are sparse) for all approximation types.
```

jlchan replied:
Changed to 

> for all `DGMulti` / `StartUpDG.jl` approximation types.

ranocha commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -503,10 +504,11 @@ end
 
-
+# calculates volume integral for <:Polynomial approximation types (e.g., we
+# do not assume any additional structure in RefElemData, such as)
```
Such as?

jlchan replied:
Added some examples

ranocha commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -598,11 +600,48 @@ function rhs!(du, u, t, mesh, equations, initial_condition, boundary_conditions:
   return nothing
 end
 
+# Specializes on SBP (e.g., nodal/collocation) DG methods with a flux differencing volume
```
Why is this required?

jlchan replied:
Because `DGMultiFluxDiffSBP  <: DGMultiFluxDiff`.

Previously, entropy projection methods had a `rhs!` which specialized on `DGMultiFluxDiff{Union{<:Polynomial, <:GaussSBP}}`, while `SBP` discretizations simply used a different volume kernel in the `rhs!` for a traditional `DGMulti` discretization.

ranocha reviewed (changes requested):
Sounds reasonable to me :+1:

jlchan commented:
@ranocha I added one more commit. I removed the `<:Polynomial` restriction from a specialization of `has_sparse_operators` in https://github.com/trixi-framework/Trixi.jl/pull/1235/commits/8d6686a41f6261d4306334b94bebd81a17038916. This overwrites a previous fallback, so I removed the previous fallback. 

This shouldn't change the behavior of any `DGMulti` solvers, since `has_sparse_operators` is appropriately specialized otherwise.
